apollo-federation/src/sources/connect/json_selection/lit_expr.rs

```diff
@@ -0,0 +1,522 @@
+                    many0(preceded(char(','), Self::parse_property)),
+                    opt(char(',')),
+                ))),
+                |properties| {
+                    let mut output = IndexMap::default();
+                    if let Some(((first_key, first_value), rest, _trailing_comma)) = properties {
+                        output.insert(first_key.to_string(), first_value);
+                        for (key, value) in rest {
+                            output.insert(key.to_string(), value);
+                        }
+                    }
+                    Self::Object(output)
+                },
+            ),
+            tuple((spaces_or_comments, char('}'), spaces_or_comments)),
+        )(input)
+    }
+
+    // LitProperty ::= Key ":" LitExpr
+    fn parse_property(input: &str) -> IResult<&str, (String, Self)> {
+        tuple((Key::parse, char(':'), Self::parse))(input)
+            .map(|(input, (key, _, value))| (input, (key.as_string(), value)))
+    }
+
+    // LitArray ::= "[" (LitExpr ("," LitExpr)* ","?)? "]"
+    fn parse_array(input: &str) -> IResult<&str, Self> {
+        delimited(
+            tuple((spaces_or_comments, char('['), spaces_or_comments)),
+            map(
+                opt(tuple((
+                    Self::parse,
+                    many0(preceded(char(','), Self::parse)),
+                    opt(char(',')),
+                ))),
+                |elements| {
+                    let mut output = vec![];
+                    if let Some((first, rest, _trailing_comma)) = elements {
+                        output.push(first);
+                        output.extend(rest);
+                    }
+                    Self::Array(output)
+                },
+            ),
+            tuple((spaces_or_comments, char(']'), spaces_or_comments)),
+        )(input)
+    }
+
+    pub(super) fn as_i64(&self) -> Option<i64> {
+        match self {
+            Self::Number(n) => n.as_i64(),
+            _ => None,
+        }
+    }
+}
+
+impl CollectVarPaths for LitExpr {
+    fn collect_var_paths(&self) -> IndexSet<&PathSelection> {
+        let mut paths = IndexSet::default();
+        match self {
+            Self::String(_) | Self::Number(_) | Self::Bool(_) | Self::Null => {}
+            Self::Object(map) => {
+                for value in map.values() {
+                    paths.extend(value.collect_var_paths());
+                }
+            }
+            Self::Array(vec) => {
+                for value in vec {
+                    paths.extend(value.collect_var_paths());
+                }
+            }
+            Self::Path(path) => {
+                paths.extend(path.collect_var_paths());
+            }
+        }
+        paths
+    }
+}
+
+impl Hash for LitExpr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match self {
+            Self::String(s) => s.hash(state),
+            Self::Number(n) => n.to_string().hash(state),
```

Comment:
Do we want the string "1" to have the same hash as the literal 1?

Reply:
Fair question, but I think I can remove the need to implement `Hash` here, so we don't have to worry about this.

Reply:
Done: https://github.com/apollographql/router/pull/5762/commits/3c925619b1d23c093d94fafda290c1ccfdaad1cc